### ISSUE TITLE
BugReport.handleCrash in updateThread hangs on currentThread().join()

### ISSUE THREAD

**Issue by werkt** [bazelbuild/bazel#22051]
### Description of the bug:

UiEventHandler's updateThread will attempt to uninterruptiblyJoin() itself when an uncaught exception occurs in conjunction with BlazeRuntime's BugReport handler. The guava join util has no detection of the deadlock that occurs when a thread joins itself.

The trace is from 6.2.0, but the lines are updated here to reflect their master position, where this is still possible:

```
"cli-update-thread" #2223 daemon prio=5 os_prio=0 cpu=357.24ms elapsed=14822.22s tid=0x00007fc8b09a2000 nid=0x3da066 in Object.wait()  [0x00007fc7452d5000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.6/Native Method)
        - waiting on <no object reference available>
        at java.lang.Thread.join(java.base@11.0.6/Unknown Source)
        - waiting to re-lock in wait() <0x00007fca0d4b0920> (a java.lang.Thread)
        at java.lang.Thread.join(java.base@11.0.6/Unknown Source)
        at com.google.common.util.concurrent.Uninterruptibles.joinUninterruptibly(Uninterruptibles.java:162)
        at com.google.devtools.build.lib.runtime.UiEventHandler.stopUpdateThread(UiEventHandler.java:964)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/runtime/UiEventHandler.java#L982
Where `threadToWaitFor == Thread.currentThread()`
```
        at com.google.devtools.build.lib.runtime.UiEventHandler.handle(UiEventHandler.java:464)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/runtime/UiEventHandler.java#L468
```
        at com.google.devtools.build.lib.events.Reporter.handle(Reporter.java:127)
```
Line matches
```
        at com.google.devtools.build.lib.bugreport.BugReport.handleCrash(BugReport.java:274)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/bugreport/BugReport.java#L297
Note that the deadlock detection above this has not triggered
```
        - locked <0x00007fc9cd132f70> (a java.lang.Object)
        at com.google.devtools.build.lib.bugreport.BugReport.handleCrash(BugReport.java:231)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/bugreport/BugReport.java#L245
```
        at com.google.devtools.build.lib.runtime.BlazeRuntime.lambda$newRuntime$4(BlazeRuntime.java:1217)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java#L1228
```
        at com.google.devtools.build.lib.runtime.BlazeRuntime$$Lambda$90/0x00007fc94a36fc40.handleException(Unknown Source)
        at com.google.devtools.build.lib.runtime.BlazeRuntime.lambda$newRuntime$5(BlazeRuntime.java:1227)
```
https://github.com/bazelbuild/bazel/blob/076494cb1ad57621cd0c883c0450569c9b3f2841/src/main/java/com/google/devtools/build/lib/runtime/BlazeRuntime.java#L1238
```
        at com.google.devtools.build.lib.runtime.BlazeRuntime$$Lambda$91/0x00007fc94a36f458.uncaughtException(Unknown Source)
        at java.lang.ThreadGroup.uncaughtException(java.base@11.0.6/Unknown Source)
        at java.lang.ThreadGroup.uncaughtException(java.base@11.0.6/Unknown Source)
        at java.lang.Thread.dispatchUncaughtException(java.base@11.0.6/Unknown Source)
```

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Throw an uncaught exception in the updateThread "cli-update-thread" with the Thread.setDefaultUncaughtExceptionHandler -> BugReport.handleCrash sequence established in BlazeRuntime.java

### Which operating system are you running Bazel on?

linux

### What is the output of `bazel info release`?

6.2.0

### Any other information, logs, or outputs that you want to share?

Original trace:

```
"cli-update-thread" #2223 daemon prio=5 os_prio=0 cpu=357.24ms elapsed=14822.22s tid=0x00007fc8b09a2000 nid=0x3da066 in Object.wait()  [0x00007fc7452d5000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@11.0.6/Native Method)
        - waiting on <no object reference available>
        at java.lang.Thread.join(java.base@11.0.6/Unknown Source)
        - waiting to re-lock in wait() <0x00007fca0d4b0920> (a java.lang.Thread)
        at java.lang.Thread.join(java.base@11.0.6/Unknown Source)
        at com.google.common.util.concurrent.Uninterruptibles.joinUninterruptibly(Uninterruptibles.java:162)
        at com.google.devtools.build.lib.runtime.UiEventHandler.stopUpdateThread(UiEventHandler.java:964)
        at com.google.devtools.build.lib.runtime.UiEventHandler.handle(UiEventHandler.java:464)
        at com.google.devtools.build.lib.events.Reporter.handle(Reporter.java:127)
        at com.google.devtools.build.lib.bugreport.BugReport.handleCrash(BugReport.java:274)
        - locked <0x00007fc9cd132f70> (a java.lang.Object)
        at com.google.devtools.build.lib.bugreport.BugReport.handleCrash(BugReport.java:231)
        at com.google.devtools.build.lib.runtime.BlazeRuntime.lambda$newRuntime$4(BlazeRuntime.java:1217)
        at com.google.devtools.build.lib.runtime.BlazeRuntime$$Lambda$90/0x00007fc94a36fc40.handleException(Unknown Source)
        at com.google.devtools.build.lib.runtime.BlazeRuntime.lambda$newRuntime$5(BlazeRuntime.java:1227)
        at com.google.devtools.build.lib.runtime.BlazeRuntime$$Lambda$91/0x00007fc94a36f458.uncaughtException(Unknown Source)
        at java.lang.ThreadGroup.uncaughtException(java.base@11.0.6/Unknown Source)
        at java.lang.ThreadGroup.uncaughtException(java.base@11.0.6/Unknown Source)
        at java.lang.Thread.dispatchUncaughtException(java.base@11.0.6/Unknown Source)
```

**Comment by tjgq** [comment]
It sounds like we could fix this by checking whether we're joining with the current thread. Could you send a PR?

**Comment by werkt** [comment]
> It sounds like we could fix this by checking whether we're joining with the current thread. Could you send a PR?

I certainly can. Hopefully there's no chance that we're stopUpdateThread-ing when in the updateThread where we aren't handling a FATAL, otherwise I don't know if it will shut down...

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.2.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.2.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.2.0rc1. Thanks!
